permission_handler_android/lib/src/android_permission_handler_api_impls.dart

```diff
@@ -0,0 +1,103 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter_instance_manager/flutter_instance_manager.dart';
+
+import 'android_object_mirrors/activity.dart';
+import 'permission_handler.pigeon.dart';
+
+/// Host API implementation of ActivityCompat.
+class ActivityCompatHostApiImpl extends ActivityCompatHostApi {
+  /// Creates a new instance of [ActivityCompatHostApiImpl].
+  ActivityCompatHostApiImpl({
+    this.binaryMessenger,
+    InstanceManager? instanceManager,
+  })  : instanceManager = instanceManager ?? JavaObject.globalInstanceManager,
+        super(binaryMessenger: binaryMessenger);
+
+  /// Sends binary data across the Flutter platform barrier.
+  ///
+  /// If it is null, the default BinaryMessenger will be used which routes to
+  /// the host platform.
+  final BinaryMessenger? binaryMessenger;
+
+  /// Maintains instances stored to communicate with native language objects.
+  final InstanceManager instanceManager;
+
+  /// Gets whether you should show UI with rationale before requesting a permission.
+  Future<bool> shouldShowRequestPermissionRationaleFromInstance(
+    Activity activity,
+    String permission,
+  ) async {
+    final String activityInstanceId = instanceManager.getIdentifier(activity)!;
+
+    return shouldShowRequestPermissionRationale(
+      activityInstanceId,
+      permission,
+    );
+  }
+}
+
+/// Flutter API implementation of Activity.
+class ActivityFlutterApiImpl extends ActivityFlutterApi {
+  /// Constructs a new instance of [ActivityFlutterApiImpl].
+  ActivityFlutterApiImpl({
+    InstanceManager? instanceManager,
+  }) : _instanceManager = instanceManager ?? JavaObject.globalInstanceManager;
+
+  /// Maintains instances stored to communicate with native language objects.
+  final InstanceManager _instanceManager;
+
+  /// The activity currently attached to the Flutter engine.
+  ///
+  /// This is null when no activity is attached.
+  Activity? _activity;
+
+  /// Registered callbacks for activity changes.
+  ///
+  /// Registered when invoking the factory constructor. These are all called
+  /// everytime an activity attaches or detaches.
+  final List<void Function(Activity? activity)> _onActivityChangedCallbacks =
+      [];
+
+  /// Registers a callback that is invoked with the current activity, everytime
+  /// it changes.
+  void registerOnActivityChanged(
+    void Function(Activity? activity) onActivityChanged,
+  ) {
+    // Emit the current attached activity.
+    onActivityChanged(_activity);
+
+    _onActivityChangedCallbacks.add(onActivityChanged);
+  }
```

Comment:
If we support registering multiple callbacks (which I think we should), we should also support removing a callback when users no longer want to listen. 

I think we should also rename the method to `addOnAttachedToActivityCallback` and create a matching method `removeOnAttachedToActivityCallback`.